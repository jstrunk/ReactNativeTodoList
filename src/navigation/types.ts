import { StackNavigationProp } from "@react-navigation/stack";
import { RouteProp } from "@react-navigation/native";

export type RootStackParamList = {
  TodoSelect: undefined;
  TodoList: {id: string};
};

export type TodoSelectScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'TodoSelect'
>;

export type TodoListScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'TodoList'
>;

export type TodoListScreenRouteProp = RouteProp<
  RootStackParamList,
  'TodoList'
>;

