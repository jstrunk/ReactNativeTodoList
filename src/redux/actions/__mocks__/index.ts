export const addItem = jest.fn();
export const toggleItem = jest.fn();
export const deleteItem = jest.fn(); 
export const deleteCompleted = jest.fn(); 
export const reorderList = jest.fn(); 
export const registerList = jest.fn(); 
export const setActiveList = jest.fn(); 
export const createList = jest.fn(); 
export const loadList = jest.fn(); 