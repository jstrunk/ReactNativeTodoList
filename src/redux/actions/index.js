import { ADD_ITEM, DELETE_ITEM, TOGGLE_ITEM, DELETE_COMPLETED } from '../constants';

export function addItem(text) {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index) {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index) {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted() {
  return { type: DELETE_COMPLETED };
}
