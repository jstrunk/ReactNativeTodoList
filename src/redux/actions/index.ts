import { ADD_ITEM,
  DELETE_ITEM,
  TOGGLE_ITEM,
  DELETE_COMPLETED,
  TodoActionType, } from '../types';

export function addItem(text: string): TodoActionType {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index: number): TodoActionType {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index: number): TodoActionType {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted(): TodoActionType {
  return { type: DELETE_COMPLETED };
}
