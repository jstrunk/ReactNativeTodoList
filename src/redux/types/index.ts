export interface Todo {
  key: string;
  done: boolean;
}

export interface ITodoList {
  byId: Record<string,Todo>;
  allItems: Array<string>;
}

export interface ITodoState {
  todoList: ITodoList;
}

export const defaultState: ITodoList = {
  byId: {},
  allItems: [],
};

export const ADD_ITEM = "ADD_ITEM";
export const TOGGLE_ITEM = "TOGGLE_ITEM";
export const DELETE_ITEM = "DELETE_ITEM";
export const DELETE_COMPLETED = "DELETE_COMPLETED";

interface AddItemAction {
  type: typeof ADD_ITEM;
  text: string;
  id: string;
}

interface ToggleItemAction {
  type: typeof TOGGLE_ITEM;
  id: string;
}

interface DeleteItemAction {
  type: typeof DELETE_ITEM;
  id: string;
}

interface DeleteCompletedAction {
  type: typeof DELETE_COMPLETED;
  ids: Array<string>;
}

export type TodoActionType = AddItemAction | ToggleItemAction |
  DeleteItemAction | DeleteCompletedAction;
