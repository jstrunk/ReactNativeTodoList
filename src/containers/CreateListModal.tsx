import React, { Component } from 'react';
import { Modal, TextInput, View } from 'react-native';
import { createList } from '../redux/actions';
import { connect, ConnectedProps } from 'react-redux';
import itemId from '../lib/itemId';
import { ITodoState, ITodoIndex } from '../redux/types';
import styles from '../styles';

interface IProps {
  isCreateListModalVisible: boolean;
  toggleCreateListModal: () => void;
}

const mapStateToProps = (state: ITodoState) => ({
  collection: state.todoCollection.listIds,
})

const MapDispatch = {
  createList: (id: string, name: string) => createList(id, name),
};

const connector = connect(
  mapStateToProps,
  MapDispatch,
);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = IProps & PropsFromRedux;

type CreateListState = {
  name: string;
}

class CreateListModal extends Component<Props, CreateListState> {
  createListInput: React.RefObject<TextInput>;
  constructor(props: Props) {
    super(props);
    this.state = {
      name: '',
    };
    this.createListInput = React.createRef<TextInput>();
  }
  onChangeText = (text: string): void => {
    this.setState({name: text});
  }

  submitItem = async () => {
    if (this.state.name == '') return;
    let newId = await itemId(this.state.name);
    if (this.props.collection.some((item: ITodoIndex) => newId === item.id)) {
      newId = await itemId(this.state.name + '1');
    }
    this.props.createList(newId, this.state.name);
    this.setState({name: ''});
    if (this.createListInput !== null && this.createListInput.current !== null) {
      this.createListInput.current.setNativeProps({text: ''});
    }
    this.props.toggleCreateListModal();
  }

  render() {
    return (
      <View style={styles.centeredView}>
        <Modal visible={this.props.isCreateListModalVisible} transparent={true} >
          <View style={styles.centeredView}>
            <View style={styles.modalView} >
              <TextInput
                ref={this.createListInput}
                style={styles.createListTextInput}
                onChangeText={this.onChangeText}
                onSubmitEditing={this.submitItem}
                placeholder='Create new list'
              />
            </View>
          </View>
        </Modal>
      </View>
    );
  }

}

export default connector(CreateListModal);