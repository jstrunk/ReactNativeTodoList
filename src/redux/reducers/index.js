import { combineReducers } from 'redux';
import listReducer from '../reducers/listReducer';

export default combineReducers(
  { todoList: listReducer }
);