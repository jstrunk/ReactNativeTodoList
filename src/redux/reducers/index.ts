import { combineReducers } from 'redux';
import listReducer from '../reducers/listReducer';
import itemReducer from '../reducers/itemReducer';
import { ITodoState } from '../types';

export default combineReducers<ITodoState>(
  {
    todoList: listReducer,
    todoItems: itemReducer,
  }
);
