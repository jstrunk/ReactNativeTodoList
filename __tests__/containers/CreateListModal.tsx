import 'react-native';
import React from 'react';
import CreateListModal from '../../src/containers/CreateListModal';
import * as types from '../../src/redux/types';
import * as actions from '../../src/redux/actions';
import { Provider } from 'react-redux';
import { render, fireEvent } from '@testing-library/react-native'
import configureMockStore, { MockStore } from 'redux-mock-store';
import thunk, { ThunkDispatch } from 'redux-thunk';
import { oneListOneItemDone } from '../../test/testStates';

jest.mock('../../src/lib/itemId');

jest.mock('../../src/redux/actions');

type DispatchExts = ThunkDispatch<types.ITodoState, void, types.CombinedActionType>;
const middlewares = [thunk];
const mockStore = configureMockStore<types.ITodoState,DispatchExts>(middlewares);

describe('CreateListModal container', () => {
  let store: MockStore;
  let toggleCreateListModal: jest.Mock<any,any>;

  beforeEach(() => {
    store = mockStore(oneListOneItemDone);

    store.dispatch = jest.fn();
    toggleCreateListModal = jest.fn();
  });

  it('renders an CreateListModal', () => {
    const visible = true;
    const { toJSON } = render(
      <Provider store={store}>
        <CreateListModal toggleCreateListModal={toggleCreateListModal} isCreateListModalVisible={visible} />
      </Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should dispatch a create list action', async () => {
    const visible = true;
    const { getByPlaceholderText } = render(
      <Provider store={store}>
        <CreateListModal toggleCreateListModal={toggleCreateListModal} isCreateListModalVisible={visible} />
      </Provider>
    );

    const element = getByPlaceholderText('Create new list');

    fireEvent.changeText(element, 'Hello World!');
    fireEvent(element, 'submitEditing');

    await new Promise((r) => setTimeout(r, 2000));

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      actions.createList('Hello World!', 'Hello World!')
    );
    expect(toggleCreateListModal).toHaveBeenCalledTimes(1);
  });

  it('should handle id collisions', async () => {
    const visible = true;
    const { getByPlaceholderText } = render(
      <Provider store={store}>
        <CreateListModal toggleCreateListModal={toggleCreateListModal} isCreateListModalVisible={visible} />
      </Provider>
    );

    const element = getByPlaceholderText('Create new list');

    fireEvent.changeText(element, 'list1');
    fireEvent(element, 'submitEditing');
  
    await new Promise((r) => setTimeout(r, 2000));

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      actions.createList('list11', 'list1')
    );
    expect(toggleCreateListModal).toHaveBeenCalledTimes(1);
  });

  it('should not dispatch on empty input', async () => {
    const visible = true;
    const { getByPlaceholderText } = render(
      <Provider store={store}>
        <CreateListModal toggleCreateListModal={toggleCreateListModal} isCreateListModalVisible={visible} />
      </Provider>
    );

    const element = getByPlaceholderText('Create new list');

    fireEvent.changeText(element, '');
    fireEvent(element, 'submitEditing');

    await new Promise((r) => setTimeout(r, 2000));

    expect(store.dispatch).toHaveBeenCalledTimes(0);
  });
});