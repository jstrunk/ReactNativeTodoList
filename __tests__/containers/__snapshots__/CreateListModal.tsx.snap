// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateListModal container renders an CreateListModal 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "marginTop": 22,
    }
  }
>
  <Modal
    hardwareAccelerated={false}
    transparent={true}
    visible={true}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginTop": 22,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "lightgrey",
            "borderRadius": 20,
            "elevation": 5,
            "margin": 20,
            "padding": 35,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 3.84,
            "width": "75%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Create new list"
          rejectResponderTermination={true}
          style={
            Object {
              "borderBottomWidth": 1,
              "padding": 15,
              "textAlign": "left",
              "width": "100%",
            }
          }
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
  </Modal>
</View>
`;
